### ISSUE TITLE
Performance of lfilter on CUDA

### ISSUE THREAD

**Issue by tqbl** [pytorch/audio#3628]
When using `lfilter()` with CUDA enabled, the time taken appears to increase by an order of magnitude after a certain number of calls to the function, which seems strange to me.

Example code:
```python
import timeit

import torch
import torchaudio


torch.set_default_device("cuda")
x = torch.rand(48000)
a = torch.tensor([1, -0.99])
b = torch.tensor([1 - 0.99, 0])

def lfilter():
    return torchaudio.functional.lfilter(x, a, b)

lfilter()  # Warmup
for i in range(150):
    # Print the number of calls so far along with the time taken.
    print(i + 2, timeit.timeit(lfilter, number=1))
```
Output (truncated):
```
110 0.00040172203443944454
111 0.0004422110505402088
112 0.0004018489271402359
113 0.0027625319780781865
114 0.006800470990128815
115 0.006813084008172154
116 0.006803544936701655
117 0.006811343948356807
118 0.006810071994550526
119 0.0068132010055705905
120 0.006845887983217835
```
From the 113th call onwards, the time taken increases considerably.

I tried both the nightly version of torchaudio and the stable version (2.0.1).

Here are the details of my environment:
[env.txt](https://github.com/pytorch/audio/files/12772812/env.txt)

**Edit**: I can't reproduce this when the length of the signal is sufficiently small, e.g. 8192.

**Comment by tqbl** [comment]
A little off-topic (probably not worth creating a new thread), but I found that lfilter on the CPU is significantly slower than scipy's implementation -- about 3x slower using the first-order filter in the code above.

Also, the docs for the function say that the input must be normalized to [-1, 1]. Isn't this only when `clamp=True`, or are there issues regardless?

**Comment by mthrok** [comment]
@yoyololicon Do you have an insight on this?

**Comment by yoyololicon** [comment]
@tqbl, thanks for opening the issue.
I can reproduce the same result on my side.
I'll find a time to look into this.

**Comment by tqbl** [comment]
@yoyololicon After revisiting this, I think the issue is with the benchmarking method rather than the function itself. If I call `torch.cuda.synchronize()` in between invocations or ensure synchronisation is done implicitly (e.g. by calling `.item()` on one of the output elements), the time taken is approximately the same across invocations.

So the lesson here is to take the asynchronicity of CUDA operations into consideration when running benchmarks. `torch.utils.benchmark.Timer` seems to be more reliable in this respect.
